draft-ietf-mls-protocol.md

```diff
@@ -1537,8 +1538,8 @@ more than one message.
 Keys, nonces, and the secrets in ratchets are derived using
 DeriveTreeSecret. The context in a given call consists of the index
 of the sender's leaf in the ratchet tree and the current position in
-the ratchet.  In particular, the index of the sender's leaf in the
-ratchet tree is the same as the index of the leaf in the Secret Tree
+the ratchet.  In particular, the node index of the sender's leaf in the
+ratchet tree is the same as the node index of the leaf in the Secret Tree
```

Comment:
This might be me being confused, however I understood the "index" in this sentence (and the subsequent key derivation) referring to N, and N being the node index:

> Nodes are also assigned an index according to their position in the array representation of the tree (described in Appendix A).

Given that we derive, e.g., `handshake_ratchet_secret_[N]_[0]` from `tree_node_[N]_secret`, and the latter being defined on these node indices, I assumed that, throughout, N should be a node index.

If you say the symmetric ratchets `..._ratchet_secret_...` are identified via _leaf indices_, then I'd agree this should all better refer to those leaf indices.

Reply:
This seems to be quite under-specified indeed. I was working under the assumption that members are always referenced by leaf indexes, because node indexes are only relevant for operation within the tree, not at the leaf level.
@Bren2010 @bifurcation any opinion on which it is?

Reply:
The `N` in the indices in subscripts and context values is a **node** index, not a **leaf** index.  You need that to be the case so that you don't have collisions from switching between them.  If you want, we could clarify that the member at leaf `n` uses `ratchet_secret_[2*n]_[*]` (recall that the node index of the `n`-th leaf is `2*n`).

Reply:
Is there any reason for keeping both node and leaf index around conceptually? I think we're only ever indexing the tree array, so why not just use node indices everywhere?

Reply:
The case that comes immediately to mind: In Remove, it's nice if the syntax constrains you so that you can only remove leaf nodes, not arbitrary nodes.

Reply:
It would be more robust if we exclusively used node indices. You still have to be mindful about where you have to convert and where you don't. We can add a simple check to Remove to make sure that only leaf nodes get removed. We have to check that the leaf is within the tree size anyway.

You told me before that we don't add stuff to the standard without good reason, so why now keep it in? :wink: 
Beside saving us the check at Remove: Is there really a reason to keep leaf indices around?

Reply:
To second what Konrad said: We need to do checks for Removes anyway, since you can't remove an empty leaf node and the index has to be <= than the number of members, so this would just be one more test.
I like the idea to have only one type of index for all operations that are covered by the spec. Naturally applications can then choose to reference members with a different scheme if they so wish (especially since empty leaf nodes are already an issue today if the member list uses leaf indexes).

Reply:
- Leaf index is used in Remove and in `unmerged_leaves`, and also in internal APIs that refer to specific members
- Like Richard, I haven't had trouble keeping leaf and node ids separate in my implementation
- I personally like keeping the two separate because it's more semantic

Reply:
I'll create a separate issue for this, so this PR can go ahead. 

I'm aware that leaf indices are used, I just don't think they are necessary. As far as I can see, in every place they are used you can simply use a node index instead, with the exception of the Remove, where you have to add a simple check. Just because you _can_ implement it (and have already done so) doesn't mean it should be in the spec.

Reply:
Thanks for following that issue, @kkohbrok.  If that issue is off the table, I think this is clear to merge.